[feature] Implement Iperf check #385

Aryamanz29 opened the pull request:
- Added initial code for **Iperf check class.**
- Added **tests** for check class.
- Added **device connection** logic.
- Added **charts & metric (TCP & UDP mode)**.

## `Steps to test`
1. Make sure your client **(openwrt-device)** and server both have [Iperf3](https://iperf.fr/iperf-download.php), [Openwrt Iperf3 package](https://openwrt.org/packages/pkgdata/iperf3)  installed.
2.  Do check **the credential section of the device** _(It must be enabled and working with the right update strategy, i.e., **ssh**)_.

![173354971-728cb1b4-4bf1-4d89-9c22-ce9b9b1e2bef](https://user-images.githubusercontent.com/56113566/173355074-3918d860-a4ad-4f84-bfa4-2c23a7e8d94e.png)

3.  In `tests/openwisp2/settings.py` configure:
```py
OPENWISP_MONITORING_IPERF_SERVERS = {
    # Running on my local
    # Some Public Iperf Servers : https://iperf.fr/iperf-servers.php#public-servers
    # 'be63c4e5-a68a-4650-bfe8-733837edb8be': ['iperf.biznetnetworks.com'],
    'a9734710-db30-46b0-a2fc-01f01046fe4f': ['speedtest.uztelecom.uz'],
    # '<org-pk>': ['<ORG_IPERF_SERVER>']
}

```
 **NOTE :** The host can be specified by **hostname, IPv4 literal, or IPv6 literal**
```
# for ex
              iperf3 -c iperf.biznetnetworks.com

              iperf3 -c 192.168.5.109

              iperf3 -c 2001:db8::1
```
4. Run `python3 manage.py run_checks` _(To run all checks manually)_.


![-------------------------------------------------------------](https://raw.githubusercontent.com/andreasbm/readme/master/assets/lines/rainbow.png)

## `Demo`

https://user-images.githubusercontent.com/56113566/173353857-56b344f4-89d2-47d9-884a-32a472168f66.mp4

![-------------------------------------------------------------](https://raw.githubusercontent.com/andreasbm/readme/master/assets/lines/rainbow.png)

## `TCP Charts`

![Screenshot from 2022-06-24 18-27-43](https://user-images.githubusercontent.com/56113566/175540716-145b7736-f8c1-4293-bfa1-9234a3ffb95a.png)

![Screenshot from 2022-06-24 11-51-01](https://user-images.githubusercontent.com/56113566/175475141-4f401c73-5198-490d-b6a1-a3932182051f.png)


## `UDP Charts`

![Screenshot from 2022-06-29 20-31-49](https://user-images.githubusercontent.com/56113566/176470082-4d94a537-9032-4f2a-b7e0-3f607ee67c5b.png)

![Screenshot from 2022-06-22 13-26-21](https://user-images.githubusercontent.com/56113566/174975685-6a101362-0132-4d88-a16f-486507f61280.png)

![Screenshot from 2022-06-22 13-26-25](https://user-images.githubusercontent.com/56113566/174975707-96a7b909-c709-4b6b-8982-33fa0e8a555d.png)

Closes #385

nemesisdesign reviewed (commented):
I changed the target branch to [gsoc22-iperf](https://github.com/openwisp/openwisp-monitoring/tree/gsoc22-iperf). Please let's use this from now on.

CC: @pandafy @okraits.

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,186 @@
+
+class Iperf(BaseCheck):
+    def check(self, store=True):
+        try:
+            device = self.related_object
+            device_connection = self._check_device_connection(device)
```
Let's only put the code that can raise exception in try-except block. I think only these lines can raise `ObjectDoesNotExist` exception here.

Aryamanz29 replied:
Thanks for the suggestion, I've already changed this in my local env, Will push soon :)

Aryamanz29 replied:
> Thanks for the suggestion, I've already changed this in my local env, Will push soon :)

Acc to https://github.com/openwisp/openwisp-monitoring/pull/394#discussion_r902404374, `_get_device_connection` method now only return either `device_connection` object or `None` so we can now safely remove `try-except` block from the code.

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,186 @@
+            chart.full_clean()
+            chart.save()
+
+    def _create_alert_settings(self, metric):
+        pass
```
What should trigger this alert? Is it going to be something like `if bandwidth falls below X Mbps, then send alert`.

Aryamanz29 replied:
We can do that as well, but according to **possible outcomes** of the project ''If an Iperf server is down/busy we should trigger an alert" (We need to investigate more about this)

pandafy commented on openwisp_monitoring/tests/test_selenium.py:
```diff
@@ -89,6 +89,8 @@ def test_restoring_deleted_device(self):
         org = self._get_org()
         self._create_credentials(auto_add=True, organization=org)
         device = self._create_config(organization=org).device
+        # Delete iperf check to prevent unnecessary response timeout
+        Check.objects.filter(check_type__endswith='Iperf').delete()
```
@Aryamanz29 can you share why this is needed?

Aryamanz29 replied:
While running tests, I'm getting `TimeoutException` due to extra check, Now it's working fine, I'll remove that line :+1: 
```py
def test_restoring_deleted_device(self):
...
...       
      except TimeoutException:
            self.fail('Failed saving device')
```

pandafy reviewed (changes requested):
Great work @Aryamanz29! 👏🏼 👏🏼 

@nemesisdesign  @okraits  things that I have tested manually: 
- The `iperf` check is created automatically when a new device registers 
- The `iperf` check is not created if `OPENWISP_MONITORING_AUTO_IPERF` is set to `False`.
- The `iperf` check execute for a device having an active SSH connection. 
- The `iperf` check does not execute for a device whose SSH connection is in `not_working` state
- The charts are shown for the gathered data. 

**Edge cases to handle**
- The database reports that `DeviceConnection` is working, but the management tunnel goes down just before `iperf` check was executed. The code should be able to gracefully handle this and also notify users that SSH connection is not working. We already have a mechanism for sending notifications. @Aryamanz29 please check other checks for reference and ask us your doubts. 

**Things to discuss in next weekly meeting**
- What parameter should trigger alert for `iperf` check? 
- Which data should be shown to users with charts?

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,186 @@
+            sent_json = res_dict['end']['sum_sent']
+            recv_json = res_dict['end']['sum_received']
+            sent_bytes = sent_json['bytes']
+            sent_bytes_GB = sent_bytes / 1000000000
```
Yes, I'm thinking of same we can leverage `traffic charts` adaptive size (https://github.com/openwisp/openwisp-monitoring/issues/87) in `Iperf` charts as well.

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,186 @@
+            if device_connection:
+                device_connection.connect()
+                servers = self._get_iperf_servers(device.organization.id)
+                command = f'iperf3 -c {servers[0]} -J'
```
Sure 👍

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,186 @@
+        try:
+            device = self.related_object
+            device_connection = self._check_device_connection(device)
+            if device_connection:
```
Thanks for the suggestion, I've already changed this in my local env, Will push soon :)

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,186 @@
+        Check device has an active connection with right update_strategy(ssh)
+        """
+        openwrt_ssh = UPDATE_STRATEGIES[0][0]
+        device_connection = DeviceConnection.objects.get(device_id=device.id)
```
You're right, we should do with .`filter` as you've mention above https://github.com/openwisp/openwisp-monitoring/pull/394#discussion_r902404374

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,186 @@
+            if (
+                device_connection.enabled
+                and device_connection.is_working
+                and device.monitoring.status in ['ok', 'problem']
```
I did this condition check to solve https://github.com/openwisp/openwisp-monitoring/issues/399, later found it is due to something else so removed in https://github.com/openwisp/openwisp-monitoring/pull/394/commits/b60c7b8408ec23f43a6769a7e71f11006f2c5f21

Aryamanz29 commented:
@pandafy  In [b535521](https://github.com/openwisp/openwisp-monitoring/pull/394/commits/b5355215a21cb60a2f857600e70b1730eeb5b130) I have fixed https://github.com/openwisp/openwisp-monitoring/issues/399, Now you can see below that error is longer produced because earlier we're executing command without considering the `device_connection` working status just right after `connect()` method.

One more thing, do we really need to **notify users** for this event?, I mean when I'm testing it locally, the _**"SSH not working"**_ alert already popped out & it already fulfilled our purpose, what do you think?  Although, I have added `logger.warn()` which is specific for the check.

![Screenshot from 2022-06-23 15-21-42](https://user-images.githubusercontent.com/56113566/175273931-c2ed5d94-3afd-4edc-aee1-aa28e3152fb0.png)

![Screenshot from 2022-06-23 15-21-46](https://user-images.githubusercontent.com/56113566/175276221-3439986e-5643-4e36-8309-07836dfac603.png)

pandafy commented:
> One more thing, do we really need to **notify users** for this event? I mean when I'm testing it locally, the _**"SSH not working"**_ alert already popped out & it already fulfilled our purpose, what do you think? Although, I have added `logger.warn()` which is specific for the check.

As the user is already getting one notification for **SSH not working**, we don't need to create another notification. Logging a warning is sufficient. 

I will do manual testing later today.

pandafy reviewed (commented):
@nemesisdesign @okraits I need some clarity on how we are planning to store the data in the InfluxDB. The current implementation ignores the bitrate from the UDP test, which seems wrong to me. 

**Thinking out loud**
A good argument could be "because UDP bitrate would be similar to TCP bitrate so we can ignore that". If so, then why bother doing a UDP test? Is it only for getting jitter and packet loss?

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,180 @@
+        command = f'iperf3 -c {servers[0]} -J'
+        res, exit_code = device_connection.connector_instance.exec_command(
+            command, raise_unexpected_exit=False
+        )
```
We can try to log the error from the output: 

```
{
        "start":        {
                "connected":    [],
                "version":      "iperf 3.10.1",
                "system_info":  "Linux openwrt-router 5.4.188 #0 Sat Apr 16 12:59:34 2022 mips"
        },
        "intervals":    [],
        "end":  {
        },
        "error":        "error - unable to connect to server: Connection refused"
}
```

Aryamanz29 replied:
Yes, it would be better to log an error here :+1:

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,180 @@
+        if not device_connection:
+            logger.warning(f'{device}: connection not properly set, Iperf skipped!')
+            return
+        device_connection.connect()
+        # We need to check device_connection is_working just right after connect().
+        # because it may be possible that authentication (publickey) failed.
+        # or any other failure happened during connect().
+        if not device_connection.is_working:
+            logger.warning(f'{device}: SSH connection is not working, Iperf skipped!')
+            return
```
I will check :+1:

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,180 @@
+            retransmits = sent_json['retransmits']
+
+            result = {
+                'iperf_result': 1,
```
I'm also thinking on this, I will do some research and try to implement different logic :+1:

Aryamanz29 replied:
Opened : https://github.com/openwisp/openwisp-monitoring/issues/405

Aryamanz29 replied:
> I'm also thinking on this, I will do some research and try to implement different logic +1

Can we do something like if any (tcp, udp) of iperf get passed, we can store `iperf_result` field to 1 otherwise 0. For eg: 
```bash
# If both test passed
'values': {'iperf_result': 1, 'sent_bps': 1.49, 'received_bps': 1.49, 'sent_bytes': 1.87, 'received_bytes': 1.87, 'retransmits': 0, 'jitter': 0.01, 'packets': 898, 'lost_packets': 0, 'lost_percent': 0.0},
# TCP passed UDP failed (previously all values was set to '0'), Now :
'values': {'iperf_result': 1, 'sent_bps': 1.5, 'received_bps': 1.5, 'sent_bytes': 1.87, 'received_bytes': 1.87, 'retransmits': 0, 'jitter': 0.0, 'packets': 0, 'lost_packets': 0, 'lost_percent': 0.0}
# Both fails
'values': {'iperf_result': 0, 'sent_bps': 0.0, 'received_bps': 0.0, 'sent_bytes': 0.0, 'received_bytes': 0.0, 'retransmits': 0, 'jitter': 0.0, 'packets': 0, 'lost_packets': 0, 'lost_percent': 0.0}
```

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,180 @@
+        org_servers = app_settings.IPERF_SERVERS.get(str(organization_id))
+        return org_servers
+
+    def _get_iperf_result(self, res, mode):
```
Initially I have also this in my mind to store the result of UDP test that includes `("bytes", "bits_per_second")` but we already calculated this in TCP test. Do we need to store these fields in UDP test as well?

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,180 @@
+            lost_percent = float(res_dict['end']['sum']['lost_percent'])
+            result = {
+                'jitter': round(jitter_ms, 2),
+                'packets': packets,
+                'lost_packets': lost_packets,
```
Same question!

Aryamanz29 commented on openwisp_monitoring/db/backends/influxdb/queries.py:
```diff
@@ -100,6 +100,51 @@
             "object_id = '{object_id}' GROUP BY time(1d)"
         )
     },
+    'bitrate': {
+        'influxdb': (
+            "SELECT MEAN(sent_bps) AS sent, "
+            "MEAN(received_bps) AS received FROM {key} WHERE "
+            "time >= '{time}' AND content_type = '{content_type}' AND "
+            "object_id = '{object_id}' GROUP BY time(1d)"
+        )
```
Done :+1:

nemesisdesign reviewed (changes requested):
@Aryamanz29 it's coming along very good, please start updating the README to mention this new check and add a dedicated section for this check which deserves some explanation.

Also make sure to maintain the test coverage stable.

Aryamanz29 commented:
> @Aryamanz29 it's coming along very good, please start updating the README to mention this new check and add a dedicated section for this check which deserves some explanation.
> 
> Also make sure to maintain the test coverage stable.

Yes, my next task after adding **alert settings** to the check would be to improve `README` and increase test coverage :+1:

nemesisdesign commented:
> > @Aryamanz29 it's coming along very good, please start updating the README to mention this new check and add a dedicated section for this check which deserves some explanation.
> > Also make sure to maintain the test coverage stable.
> 
> Yes, my next task after adding **alert settings** to the check would be to improve `README` and increase test coverage +1

Docs and tests are not something we do in phases. We do this iteratively, so every time we add new important features we add docs and tests for them, then as we improve and change the new features, we update the docs and the tests.

This is a requirement for this PR to get merged, so please proceed with this as soon as possible, do not leave it for the end.
**Docs and tests are not something we do at the end, it's iterative, keep in mind!**

pandafy commented on README.rst:
```diff
@@ -1001,6 +1116,18 @@ validating custom parameters of a ``Check`` object.
+| **default**: | ``False``   |
++--------------+-------------+
+
+Whether `Iperf <#iperf-1>`_ checks are created automatically for devices. The devices need to have SSH access from OpenWISP & must installed
+`iperf3 openwrt package <https://openwrt.org/packages/pkgdata/iperf3>`_ on devices in order for this check to work.
```
```suggestion
Whether `Iperf <#iperf-1>`_ checks are created automatically for devices. The devices must have ``SSH`` `access credential <link to access credentials in controller module>`_  & 
`iperf3 openwrt package <https://openwrt.org/packages/pkgdata/iperf3>`_ must be installed on the devices in order for this check to work.
```

pandafy commented on README.rst:
```diff
@@ -802,6 +802,61 @@ Mobile Access Technology in use
+Iperf
+~~~~~
+
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
@Aryamanz29 don't make these tables too wide. As you can observe here it is difficult to read. Please break the content into multiple lines.

pandafy commented on README.rst:
```diff
@@ -802,6 +802,61 @@ Mobile Access Technology in use
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| **measurement**:   | ``iperf``                                                                                                                                                                                                                      |
++--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| **types**:         | ``int`` (iperf_result, retransmits, total_packets, lost_packets), ``float`` (sent_bps_tcp, received_bps_tcp, sent_bytes_tcp, received_bytes_tcp, sent_bps_udp, sent_bytes_udp, jitter, lost_percent)                           |
```
Especially here, I would break it into two more more lines, one for `int` and other for `float`.

Aryamanz29 replied:
I'll update :+1:

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,205 @@
+        """
+        return dc.connector_instance.exec_command(command, raise_unexpected_exit=False)
+
+    def _connect(self, dc):
+        """
+        Connects device returns its working status
+        """
+        return dc.connect()
```
**Nitpick**: Why to use a helper function here?

Aryamanz29 replied:
Will remove this :+1:  I'm just testing something :)

Aryamanz29 replied:
I used this to make `deviceconnection.connect()` easier to mock inside `test_iperf.py`.

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,205 @@
+                sent_json = res_dict['end']['sum_sent']
+                recv_json = res_dict['end']['sum_received']
+                sent_bytes = sent_json['bytes']
+                sent_bytes_GB = sent_bytes / 1000000000
+                sent_bps = sent_json['bits_per_second']
+                sent_Gbps = sent_bps / 1000000000
+                received_bytes = recv_json['bytes']
+                received_bytes_GB = received_bytes / 1000000000
+                received_bps = recv_json['bits_per_second']
+                received_Gbps = received_bps / 1000000000
```
Didn't we decide to not do this conversion here? Please correct me if I am wrong.

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,205 @@
+                }
+            else:
+                sent_bytes = res_dict['end']['sum']['bytes']
+                sent_bytes_MB = sent_bytes / 1000000
+                sent_bps = res_dict['end']['sum']['bytes']
+                sent_Mbps = sent_bps / 1000000
```
Why not in Gbps?

Aryamanz29 replied:
It always comes out very small value when I changed to `Gbps`. We'll remove these manual conversions once https://github.com/openwisp/openwisp-monitoring/pull/397 get merged :+1:

pandafy commented on openwisp_monitoring/check/settings.py:
```diff
@@ -5,9 +5,13 @@
 AUTO_CONFIG_CHECK = get_settings_value('AUTO_DEVICE_CONFIG_CHECK', True)
 MANAGEMENT_IP_ONLY = get_settings_value('MANAGEMENT_IP_ONLY', True)
 PING_CHECK_CONFIG = get_settings_value('PING_CHECK_CONFIG', {})
+# By default it should be disabled.
+AUTO_IPERF = get_settings_value('AUTO_IPERF', True)
```
Is this a reminder? We should flip  it back to False, and set this value in settings.py

Aryamanz29 replied:
Yes it should be `false`.

pandafy commented on openwisp_monitoring/check/tests/test_iperf.py:
```diff
@@ -0,0 +1,74 @@
+        device = self._create_device()
+        self._create_config(device=device)
+        dc = self._create_device_connection(device=device, credentials=ckey)
+        check = Check(
```
This device should automatically get an iperf check according to the settings.

pandafy commented on openwisp_monitoring/check/tests/test_iperf.py:
```diff
@@ -0,0 +1,74 @@
+            check_type=self._IPERF,
+            content_object=device,
+        )
+        with self.subTest('Test inactive or invalid device connection'):
+            check.perform_check(store=False)
+            mock_warn.assert_called_with(
+                f'DeviceConnection is not properly set for "{device}", iperf check skipped!'
+            )
```
I would refactor this test in following ways to make it more readable:

1. change the subTest string to `Test device connection not working`
2. set `dc.is_working` to `False` inside the subTest

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,205 @@
+        # The DeviceConnection could fail if the management tunnel is down.
+        if not self._connect(device_connection):
+            logger.warning(
+                f'Failed to get a working DeviceConnection for "{device}", iperf check skipped!'
```
```suggestion
                f'DeviceConnection for "{device}" is not working, iperf check skipped!'
```

pandafy reviewed (changes requested):
Good progress with the documentation @Aryamanz29! I have suggested some changes below. We surely need more test. Please add tests for all the scenarios that we shared with you earlier on how the `iperf` check could fail.

pandafy commented on README.rst:
```diff
@@ -918,6 +975,69 @@ configuration status of a device changes, this ensures the check reacts
+2. Enable secure SSH access from OpenWISP to your devices
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We have already well covered this in `How to configure push updates section of openwisp-controller <https://github.com/openwisp/openwisp-controller#how-to-configure-push-updates>`_.
```
Always use relative  hyperlinks 

```suggestion
 Follow the steps in `"How to configure push updates" section of  the openwisp-controller documentation <https://github.com/openwisp/openwisp-controller#how-to-configure-push-updates>`_ to allow SSH access to 
 you device from OpenWISP.
```

pandafy commented on README.rst:
```diff
@@ -918,6 +975,69 @@ configuration status of a device changes, this ensures the check reacts
+
+We have already well covered this in `How to configure push updates section of openwisp-controller <https://github.com/openwisp/openwisp-controller#how-to-configure-push-updates>`_.
+
+**Note:** Make sure device connection is enabled & working with right update strategy ie. ``OpenWISP SSH``.
```
Isn't it **OpenWRT SSH** ?

```suggestion
**Note:** Make sure device connection is enabled & working with right update strategy i.e. ``OpenWRT SSH``.
```

Aryamanz29 replied:
Yes :+1:

pandafy commented on README.rst:
```diff
@@ -918,6 +975,69 @@ configuration status of a device changes, this ensures the check reacts
+
+Configure iperf servers in `openwisp settings <https://github.com/openwisp/openwisp-monitoring/blob/master/tests/openwisp2/settings.py>`_ , The host can be specified by hostname, IPv4 literal, or IPv6 literal.
+
+For eg.
```
```suggestion
For example:
```

pandafy commented on README.rst:
```diff
@@ -918,6 +975,69 @@ configuration status of a device changes, this ensures the check reacts
+    'c9734710-db30-46b0-a2fc-01f01046fe4f': ['2001:db8::1'],
+    }
+
+4. Run the check
```
We should also point it out here to configure the celery beat.

Aryamanz29 replied:
Noted

pandafy commented on README.rst:
```diff
@@ -1001,6 +1121,59 @@ validating custom parameters of a ``Check`` object.
+| **default**: | ``False``   |
++--------------+-------------+
+
+Whether `Iperf <#iperf-1>`_ checks are created automatically for devices. The devices must have ``SSH`` `access credential <https://github.com/openwisp/openwisp-controller#save-ssh-private-key-in-openwisp-access-credentials>`_  & 
```
I don't think we need to point out here that the device needs a working SSH connection here. 

Something like this would work fine 

```
This setting allows you to choose whether `iperf <#iperf-1>`_ checks should be
created automatically for newly registered devices. It's disabled by default.
```

pandafy commented on README.rst:
```diff
@@ -1001,6 +1121,59 @@ validating custom parameters of a ``Check`` object.
+        },
+    }
+
+Similar to `OPENWISP_MONITORING_PING_CHECK_CONFIG <#OPENWISP_MONITORING_PING_CHECK_CONFIG>`_ , 
```
Let's remove reference to the configuration of Ping check. We will make users read the section even though we are explaining everything here.

pandafy commented on README.rst:
```diff
@@ -918,6 +975,69 @@ configuration status of a device changes, this ensures the check reacts
+You can also change the default values used for iperf checks using
+`OPENWISP_MONITORING_IPERF_CHECK_CONFIG <#OPENWISP_MONITORING_IPERF_CHECK_CONFIG>`_ setting.
+
+Instructions to configure Iperf Check
```
I would create a heading `Usage Instructions` and move this section under the sub-heading `How to configure iperf check`.  This is make it similar to REAMDE of openwisp-controller. 

Wait for @nemesisdesign opinion before going ahead with this change.

nemesisdesign replied:
I am not sure I understood correctly but if you look at the result generated at https://github.com/openwisp/openwisp-monitoring/tree/issue-385/iperf-check#iperf-1, it doesn't look good (this heading is bigger than the one above), so anything better than that will be ok for me, I am pretty sure we have to give it a last round of refinement anyway @pandafy.

Aryamanz29 replied:
Improved docs :+1:

pandafy reviewed (commented):
I found an issue while deploying this. The `run_check` task is executed every 5 minutes which triggers the iperf check every 5 minutes. We should update the database query of `run_check` task to exclude iperf check.

Aryamanz29 commented:
> @Aryamanz29 what if a device doesn't have iperf installed?
I'll double check this case :+1:

Aryamanz29 commented on README.rst:
```diff
@@ -802,6 +802,63 @@ Mobile Access Technology in use
+Iperf
+~~~~~
+
++--------------------+----------------------------------------------------------------------------------------------------------------------------------------+
```
After adding `|` to my table, it looks like this : 

![Screenshot from 2022-07-08 01-23-35](https://user-images.githubusercontent.com/56113566/177860603-acd2bc75-cd62-4ad4-b338-a208073e2b76.png)

Aryamanz29 commented on README.rst:
```diff
@@ -918,6 +975,69 @@ configuration status of a device changes, this ensures the check reacts
+Iperf
+~~~~~
+
+This check provides network performance measurements such as maximum achievable bandwidth, jitter, datagram loss etc of the device using `iperf3 utility <https://iperf.fr/>`_. 
```
Got it :+1:

Aryamanz29 commented:
@nemesisdesign I have configured celery beat for iperf check as follows:
- As previous, we can still use management command `run_checks` to run all the checks in development.
- According to settings of `(CELERY_BEAT_SCHEDULE)`, Only those checks will run by `celery_beat` that are passed in `args` of celery beat scheduler configuration.

**For example (current configuration):**

```py
OPENWISP_MONITORING_CHECKS = [
    'openwisp_monitoring.check.classes.Ping',
    'openwisp_monitoring.check.classes.ConfigApplied',
    'openwisp_monitoring.check.classes.Iperf',
]

CELERY_BEAT_SCHEDULE = {
    'run_checks': {
        'task': 'openwisp_monitoring.check.tasks.run_checks',
        'schedule': timedelta(minutes=5),
        # Executes only ping and config check every 5 mins
        'args': (OPENWISP_MONITORING_CHECKS[:2],),
        'relative': True,
    },
    'run_iperf_checks': {
        'task': 'openwisp_monitoring.check.tasks.run_checks',
        # https://docs.celeryq.dev/en/latest/userguide/periodic-tasks.html#crontab-schedules
        # Executes only iperf check every 5 mins from 00:00 AM to 6:00 AM (night)
        'schedule': crontab(minute='*/5', hour='0-6'),
        'args': (OPENWISP_MONITORING_CHECKS[2:],),
        'relative': True,
    },
}
```
- Only **ping** and **config** check will be scheduled by celery beat in every **5 mins** and once we hit scheduled time for **iperf_check** then only **all three** check will start executing according to the configuration.

![Screenshot from 2022-07-09 00-25-50](https://user-images.githubusercontent.com/56113566/178056090-2328204b-ff60-4b86-af95-f958f027f530.png)

nemesisdesign commented on openwisp_monitoring/check/settings.py:
```diff
@@ -5,9 +5,13 @@
 PING_CHECK_CONFIG = get_settings_value('PING_CHECK_CONFIG', {})
+AUTO_IPERF = get_settings_value('AUTO_IPERF', False)
+IPERF_SERVERS = get_settings_value('IPERF_SERVERS', {})
+IPERF_CHECK_CONFIG = get_settings_value('IPERF_CHECK_CONFIG', {})
```
`OPENWISP_MONITORING_CHECKS` should go here and be inferred from `CHECK_CLASSES` automatically, eg:

```python
CHECKS_LIST = list(dict(CHECK_CLASSES).keys())
```

Aryamanz29 replied:
- Added `CHECKS_LIST` in `checks/settings.py` and remained same celery beat config for **ping** and **config applied** check. 
- Added instructions in  `How to configure iperf check` to configure celery beat for **iperf** check.

Aryamanz29 commented on openwisp_monitoring/check/settings.py:
```diff
@@ -15,3 +15,4 @@
 AUTO_IPERF = get_settings_value('AUTO_IPERF', False)
 IPERF_SERVERS = get_settings_value('IPERF_SERVERS', {})
 IPERF_CHECK_CONFIG = get_settings_value('IPERF_CHECK_CONFIG', {})
+CHECKS_LIST = get_settings_value('CHECK_LIST', list(dict(CHECK_CLASSES).keys()))
```
Alternate option  : 
```py
CHECKS_DICT = get_settings_value('CHECK_DICT', dict((check[::-1] for check in CHECK_CLASSES)))
# {'Ping': 'openwisp_monitoring.check.classes.Ping', 'Configuration Applied': 'openwisp_monitoring.check.classes.ConfigApplied', 'Iperf': 'openwisp_monitoring.check.classes.Iperf'}
```

pandafy reviewed (changes requested):
Good work @Aryamanz29! I have tested changing iperf check parameters from the admin dashboard as well as configuring the `OPENWISP_MONITORING_IPERF_CHECK_CONFIG` setting. 

The code works as expected in both the scenarios. There few minor things which can be improved.

pandafy commented on openwisp_monitoring/check/tasks.py:
```diff
@@ -15,17 +17,27 @@ def get_check_model():
 
 @shared_task
-def run_checks():
+def run_checks(checks=CHECKS_LIST):
```
Why don't we set this to `None` here?

Aryamanz29 replied:
Initially I did this, but what if we run management command `run_checks`, In that case it will only execute **ping** and **config applied** check instead of executing all checks in the development.

pandafy replied:
We can modify the management command to pass all three checks in the arguments. 

Running all three checks is not a standard case here. Running `ping` and `config` checks is the standard case. 

If you want to provide configurability to users, you can additionally allow passing the check dotted path as arguments to the run_checks command as follows: 

```
./manage.py run_checks openwisp_monitoring.check.classes.Iperf,openwisp_monitoring.check.classes.ConfigApplied
```

But, this is not our goal now.

pandafy commented on openwisp_monitoring/check/tasks.py:
```diff
@@ -15,17 +17,27 @@ def get_check_model():
+        # Executes only auto checks ie. ping and config applied
+        checks = CHECKS_LIST[:2]
+
+    # Checks that are included celery beat configuration
```
I think this comment is outdated.

Aryamanz29 commented on README.rst:
```diff
@@ -371,16 +372,17 @@ Configure celery (you may use a different broker if you want):
         'run_checks': {
             'task': 'openwisp_monitoring.check.tasks.run_checks',
             'schedule': timedelta(minutes=5),
-        },
-        # Delete old WifiSession
-        'delete_wifi_clients_and_sessions': {
-            'task': 'openwisp_monitoring.monitoring.tasks.delete_wifi_clients_and_sessions',
-            'schedule': timedelta(days=180),
```
By mistake 😅 I'll add :+1:

Aryamanz29 commented on openwisp_monitoring/check/tests/test_iperf.py:
```diff
@@ -0,0 +1,477 @@
+                )
+
+        with self.subTest('Test iperf check passes in both TCP & UDP'):
+            with patch.object(
+                Iperf, '_exec_command'
+            ) as mock_exec_command, patch.object(
+                Iperf,
+                '_get_iperf_servers',
+                return_value=['iperf.openwisptestserver.com'],
+            ) as mock_get_iperf_servers, patch.object(
+                iperf_logger, 'warning'
+            ) as mock_warn:
```
I did this because if I use decorator style mocks for `test_iperf_check` (which consists of various subTests) then for every subTests I'm getting additive `mock_method` call counts (It includes method calls from previous subTest) which I thought is not a good idea.
**For eg:**
```py
    @patch.object(Iperf, '_exec_command')
    @patch.object(Iperf, '_get_iperf_servers', return_value=['iperf.openwisptestserver.com'])
    @patch.object(iperf_logger, 'warning')
    def test_iperf_check(self, mock_warn, mock_get_iperf_servers, mock_exec_command):
        check, _ = self._create_iperf_test_env()
        error = "ash: iperf3: not found"
        tcp_result = loads(RESULT_TCP)['end']
        udp_result = loads(RESULT_UDP)['end']['sum']

        with self.subTest('Test iperf3 is not installed on the device'):
                mock_exec_command.side_effect = [(error, 127)]
                with patch.object(iperf_logger, 'warning') as mock_warn:
                    check.perform_check(store=False)
                    mock_warn.assert_called_with(
                        f'Iperf3 is not installed on the "{self.device}", error - {error}'
                    )
                self.assertEqual(mock_warn.call_count, 1)
                self.assertEqual(mock_exec_command.call_count, 1)
                mock_get_iperf_servers.assert_called_once_with(
                    self.device.organization.id
                )
        with self.subTest('Test iperf check passes in both TCP & UDP'):
                mock_exec_command.side_effect = [(RESULT_TCP, 0), (RESULT_UDP, 0)]
                ...
                ...

                self.assertEqual(mock_warn.call_count, 0)
                # AssertionError: 1 != 0 (Includes call count from previous subTest)
                self.assertEqual(mock_exec_command.call_count, 2)
                # AssertionError: 3 != 2 (Same here)
    
```

Aryamanz29 commented:
@pandafy Today, I've done some testing ([a273e90](https://github.com/openwisp/openwisp-monitoring/pull/394/commits/a273e90ab664a1fae2fc1a0a18da33f53c02c82d)) on alerts and notification for iperf check on `iperf_result` field. Indeed, it works according to the **alert_settings** defined in `configuration.py`, but the issue is I'm unable to see any **alert notifications (problem/recovery)** when I write data to the timeseries database below the threshold value or normal value i.e. from 0 to 1.

![Screenshot from 2022-07-13 18-21-20](https://user-images.githubusercontent.com/56113566/178737900-6f219d5b-cd1d-4211-85e1-40dad6df6ff7.png)

### Only device health changes (No notification)

![Screenshot from 2022-07-13 17-55-56](https://user-images.githubusercontent.com/56113566/178735880-b2957329-f921-484a-9472-cce80ef57e95.png)

![Screenshot from 2022-07-13 17-56-10](https://user-images.githubusercontent.com/56113566/178737641-84d01e57-80e7-4f85-bf52-ffbe714be4e6.png)

![Screenshot from 2022-07-13 17-55-55](https://user-images.githubusercontent.com/56113566/178735664-0ecafe4f-e796-4418-bfec-e2c7823071a4.png)

![Screenshot from 2022-07-13 17-56-14](https://user-images.githubusercontent.com/56113566/178738066-17e167ac-e8cf-478c-8d19-530567e33072.png)

pandafy commented:
@Aryamanz29 No AlertSetting object is present, hence no notification is created 

```
======================================================================
FAIL: test_iperf_check_alert_notification (openwisp_monitoring.check.tests.test_iperf.TestIperf)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/mock.py", line 1325, in patched
    return func(*newargs, **newkeywargs)
  File "/home/pandafy/openwisp/openwisp-monitoring/openwisp_monitoring/check/tests/test_iperf.py", line 495, in test_iperf_check_alert_notification
    self.assertEqual(AlertSettings.objects.count(), 1)
AssertionError: 0 != 1

----------------------------------------------------------------------
```

Read the tests in https://github.com/openwisp/openwisp-monitoring/blob/master/openwisp_monitoring/monitoring/tests/test_monitoring_notifications.py for reference.

Aryamanz29 commented:
> @Aryamanz29 No AlertSetting object is present, hence no notification is created
> 
> ```
> ======================================================================
> FAIL: test_iperf_check_alert_notification (openwisp_monitoring.check.tests.test_iperf.TestIperf)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/usr/lib/python3.8/unittest/mock.py", line 1325, in patched
>     return func(*newargs, **newkeywargs)
>   File "/home/pandafy/openwisp/openwisp-monitoring/openwisp_monitoring/check/tests/test_iperf.py", line 495, in test_iperf_check_alert_notification
>     self.assertEqual(AlertSettings.objects.count(), 1)
> AssertionError: 0 != 1
> 
> ----------------------------------------------------------------------
> ```
> 
> Read the tests in https://github.com/openwisp/openwisp-monitoring/blob/master/openwisp_monitoring/monitoring/tests/test_monitoring_notifications.py for reference.

Yea just noticed fixing now

Aryamanz29 commented:
### Alerts on `related_field` (Iperf jitter)

https://user-images.githubusercontent.com/56113566/178989966-f007a9b9-c9fd-4877-ac01-539e38b3331f.mp4

Aryamanz29 commented on README.rst:
```diff
@@ -1800,6 +1800,12 @@ The ``AlertSettings`` of ``ping`` metric will by default use ``threshold`` and `
+You can also use ``alert_on_related_field`` key in metric configuration
+which allows ``AlertSettings`` to use ``related_field`` as value to check ``threshold`` instead
+of default ``field_name`` key. A real world example of this can be seen in 
+`Iperf metric configuration <https://github.com/openwisp/openwisp-monitoring/blob/issue-385/iperf-check/openwisp_monitoring/monitoring/configuration.py#L546-L700>`_,
```
`Todo `: Change this link after PR merge

nemesisdesign reviewed (changes requested):
Good progress @Aryamanz29, I reviewed and tested the code with @pandafy right now.

The testing on the demo yielded weird results, the check is not really working well but right now we are not sure whether it's a misconfiguration or a series of bug, so @pandafy will double check asap on his local system.

In the meanwhile you can take a look at my comments below.

Another thing: I have noticed the results in inflxudb are not saved as bytes, was this already fixed or not?

Aryamanz29 commented on README.rst:
```diff
@@ -918,6 +984,96 @@ configuration status of a device changes, this ensures the check reacts
+1. Register your device to OpenWISP
+###################################
+
+Register your device to OpenWISP and make sure `iperf3 openwrt package <https://openwrt.org/packages/pkgdata/iperf3>`_ is installed on the device if not run :
```
👍

Aryamanz29 commented on README.rst:
```diff
@@ -802,6 +810,64 @@ Mobile Access Technology in use
+
+**Bandwidth (TCP)**:
+
+.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/bandwidth-tcp.png
```
Sure :+1:

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,251 @@
+            device_id=device.id,
+            update_strategy=openwrt_ssh,
+            enabled=True,
+            is_working=True,
```
You're right, I'll remove `is_working` from filter :+1:

Aryamanz29 commented on openwisp_monitoring/check/tests/iperf_test_result.py:
```diff
@@ -0,0 +1,662 @@
+# flake8: noqa
+
+RESULT_TCP = """{
```
Done :+1:

pandafy commented:
> The testing on the demo yielded weird results, the check is not really working well but right now we are not sure whether it's a misconfiguration or a series of bug, so @pandafy will double check asap on his local system.

I debugged the deployment of this work on the demo VM and found out this error in celery logs 

```
[2022-07-18 08:18:13,430: INFO/MainProcess] Task openwisp_monitoring.monitoring.tasks.timeseries_write[1ec11ddd-93c2-4f15-a2cd-7a9c7f3d46b4] received
[2022-07-18 08:18:25,445: WARNING/ForkPoolWorker-1] got exception while writing to tsdb: 400: {"error":"partial write: field type conflict: input field \"received_bytes_tcp\" on measurement \"iperf\" is type integer, already exists as type float dropped=1"
```

I think this occurred because earlier we were saving floating point values for sent and received bits (we were converting into Gbps)  when saving the values in the InfluxDB. Later, we changed the code to store data in `bps`. After fixing this issue at the database level, the results from iperf check were getting stored in the InfluxDB. 

The latest version of this PR is deployed on the demo system.

Aryamanz29 commented:
## `Iperf3 Auth`

### Check parameters :
- `username` : username for authentication.
- `password` : Used as **IPERF3_PASSWORD** environment variable in test command to skip password prompt.
- `rsa-public-key-path` : Path to the RSA public key used to encrypt authentication credentials.

### Template :
- `Iperf Auth template (Configuration : files)` : `/root/public.pem`: public key is used to encrypt the authentication token containing the user credentials.

### Demo

https://user-images.githubusercontent.com/56113566/179610059-f7bf6d71-2968-4dce-903c-768cd6c61579.mp4

pandafy reviewed (commented):
@Aryamanz29 I tried to test your work locally and the authentication kept failing on my local environment. 

```
Server listening on 5201
-----------------------------------------------------------
Authentication failed for user 'mario' ts 1658253759
iperf3: error - test authorization failed
```

I was following the steps that you have mentioned in https://github.com/openwisp/openwisp-monitoring/issues/414 for configuring the server. I configured the settings in OpenWISP as follows: 

```python
OPENWISP_MONITORING_IPERF_CHECK_CONFIG = {
    'username': {'default': 'mario'},
    'password': {'default':  'mariorossi'},
    'rsa-public-key-path': {'default': '/tmp/iperf-rsa.pem'}
}
```

I manually copied the contents of `public.pem` on the device. I used the following command to run the server 

```
iperf3 -s --rsa-private-key-path ./private_not_protected.pem --authorized-users-path ./credentials.csv
```

I will test it again tomorrow and double check if I made any mistake while testing this. If you find any error in settings or command, please point them out.

pandafy reviewed (changes requested):
@Aryamanz29 we need to automate the step for transferring the RSA public key to the device. I have mentioned the details on the issue https://github.com/openwisp/openwisp-monitoring/issues/414#issuecomment-1189408176.

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,269 @@
+AlertSettings = load_model('monitoring', 'AlertSettings')
+DeviceConnection = load_model('connection', 'DeviceConnection')
+
+DEFAULT_IPERF_CHECK_CONFIG = {
+    'port': {
+        'type': 'integer',
+        'default': 5201,
+        # max, min port chosen from iperf3 docs
+        'minimum': 1,
+        'maximum': 65535,
+    },
+    'time': {
+        'type': 'integer',
+        # Sets the interval time in seconds
+        # between periodic bandwidth, jitter, and loss reports.
+        'default': 10,
+        'minimum': 1,
+        # arbitrary chosen to avoid slowing down the queue (30min)
+        'maximum': 1800,
+    },
+    'username': {'type': 'string', 'default': '', 'minLength': 1, 'maxLength': 20},
+    'password': {'type': 'string', 'default': '', 'minLength': 1, 'maxLength': 20},
+    'rsa-public-key-path': {
+        'type': 'string',
+        'default': '',
+        'pattern': '^(.*/)([^/]*)$',
+    },
+}
```
I think we should set bitrate on UDP to default to 10 MB and recommend users to tweak this according to their needs.
On TCP bitrate can default to 0 (unlimited).

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -37,6 +37,13 @@
     },
+    'username': {'type': 'string', 'default': '', 'minLength': 1, 'maxLength': 20},
+    'password': {'type': 'string', 'default': '', 'minLength': 1, 'maxLength': 20},
+    'rsa-public-key-path': {
```
```suggestion
    'rsa_public_key': {
```

pandafy commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -84,23 +97,29 @@ def check(self, store=True):
-        # Exit code 127 : command doesn't exist
+        # TCP mode
+        result, exit_code = self._exec_command(device_connection, command_tcp)
+        # # Exit code 127 : command doesn't exist
```
```suggestion
        # Exit code 127 : command doesn't exist
```

pandafy commented on openwisp_monitoring/check/tests/test_iperf.py:
```diff
@@ -469,3 +527,143 @@ def test_iperf_check(self):
                     self.device.organization.id
                 )
+
+    def test_iperf_check_auth_config(self):
```
I would declare the common mocks with decorators to reduce code duplicity. 
```suggestion
    @patch.object(Iperf, '_get_iperf_servers', return_value=['iperf.openwisptestserver.com'])
    @patch.object(Iperf, '_exec_command', side_effect=[(RESULT_TCP, 0), (RESULT_UDP, 0)])
    @patch.object(iperf_logger, 'warning')
    def test_iperf_check_auth_config(self, mock_warn, mock_exec_command, mock_get_iperf_servers):
```

pandafy reviewed (changes requested):
Good work @Aryamanz29! I tested the iperf check again today and the authentication was working as expected. In the my previous attempt, the hash stored in credentials.csv was incorrect. 

Please keep improving the PR as mentioned in my previous reviews. I have left more improvements for you below.

Aryamanz29 commented on openwisp_monitoring/check/tests/test_iperf.py:
```diff
@@ -469,3 +527,143 @@ def test_iperf_check(self):
+                mock_get_iperf_servers.assert_called_once_with(
+                    self.device.organization.id
+                )
+
```
Thanks !

nemesisdesign commented on README.rst:
```diff
@@ -372,11 +373,22 @@ Configure celery (you may use a different broker if you want):
-            'task': 'openwisp_monitoring.check.tasks.run_checks',
-            'schedule': timedelta(minutes=5),
+       'run_checks': {
+        'task': 'openwisp_monitoring.check.tasks.run_checks',
```
the indentation is not right here

Aryamanz29 replied:
Done 👍

nemesisdesign commented on README.rst:
```diff
@@ -372,11 +373,22 @@ Configure celery (you may use a different broker if you want):
+                'openwisp_monitoring.check.classes.ConfigApplied',
+            ],
+        ),
+        'relative': True,
```
aren't we missing the instructions to run the iperf check?

Aryamanz29 replied:
I think it is mention [`here`](https://github.com/openwisp/openwisp-monitoring/tree/issue-385/iperf-check#3-configure-iperf-settings)

nemesisdesign commented on README.rst:
```diff
@@ -939,6 +1009,194 @@ configuration status of a device changes, this ensures the check reacts
+  :alt: Iperf network measurement charts
+
+
+Configure Iperf check for authentication
```
```suggestion
Iperf authentication
```

nemesisdesign commented on README.rst:
```diff
@@ -939,6 +1009,194 @@ configuration status of a device changes, this ensures the check reacts
+Configure Iperf check for authentication
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By default iperf check runs without any kind of **authentication**, But we can configure it to use **RSA authentication** 
+between the **client** and the **server** to restrict the connections to the server & only allow legitimate clients.
```
```suggestion
By default iperf check runs without any kind of **authentication**,
in this section we will explain how to configure **RSA authentication** 
between the **client** and the **server** to restrict connections
to authenticated clients.
```

nemesisdesign commented on README.rst:
```diff
@@ -939,6 +1009,194 @@ configuration status of a device changes, this ensures the check reacts
+   openssl rsa -in private.pem -outform PEM -pubout -out public.pem
+   openssl rsa -in private.pem -out private_not_protected.pem -outform PEM
+
+After running above mention commands, the public key will be contained in the 
```
```suggestion
After running the commands mentioned above, the public key will be stored in ``public.pem`` 
```

Keep it short, apply a similar change also to ``private_not_protected.pem``, I would also rename that as `private_key.pem`, while I would rename `public.pem` to `public_key.pem`.

Aryamanz29 replied:
Sure :)

nemesisdesign commented on README.rst:
```diff
@@ -939,6 +1009,194 @@ configuration status of a device changes, this ensures the check reacts
+.. code-block:: python
+
+    OPENWISP_MONITORING_IPERF_CHECK_CONFIG = {
+    # All three parameters are required
```
indentation is wrong here

nemesisdesign commented on README.rst:
```diff
@@ -939,6 +1009,194 @@ configuration status of a device changes, this ensures the check reacts
+    OPENWISP_MONITORING_IPERF_CHECK_CONFIG = {
+    # All three parameters are required
+    'username': {'default': 'iperfuser'},
+    'password': {'default': 'iperfpass'},
```
what is `default`?

Aryamanz29 replied:
I made it to look consistent with [`ping config override setting`](https://github.com/openwisp/openwisp-monitoring#openwisp_monitoring_ping_check_config), Now I have changed `iperf config setting` according to https://github.com/openwisp/openwisp-monitoring/issues/418

nemesisdesign commented on README.rst:
```diff
@@ -939,6 +1009,194 @@ configuration status of a device changes, this ensures the check reacts
+        BQIDAQAB
+        """
+    },
+   }
```
We already have a setting to configure iperf, why are you adding a new one?

Can we compress everything in one setting?

Eg:

```python
OPENWISP_MONITORING_IPERF = {
    'a9734710-db30-46b0-a2fc-01f01046fe4f': [
        {
            'host': ['iperf1.openwisp.io', 'iperf2.openwisp.io'],
            'username': 'iperfuser',
            'password': 'iperfpassword',
            'rsa_public_key': (
"""
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAwuEm+iYrfSWJOupy6X3N 
dxZvUCxvmoL3uoGAs0O0Y32unUQrwcTIxudy38JSuCccD+k2Rf8S4WuZSiTxaoea 
6Du99YQGVZeY67uJ21SWFqWU+w6ONUj3TrNNWoICN7BXGLE2BbSBz9YaXefE3aqw 
GhEjQz364Itwm425vHn2MntSp0weWb4hUCjQUyyooRXPrFUGBOuY+VvAvMyAG4Uk 
msapnWnBSxXt7Tbb++A5XbOMdM2mwNYDEtkD5ksC/x3EVBrI9FvENsH9+u/8J9Mf 
2oPl4MnlCMY86MQypkeUn7eVWfDnseNky7TyC0/IgCXve/iaydCCFdkjyo1MTAA4 
BQIDAQAB
""")
            'client_options': {
                'port': '6209',
                'udp': {
                    'bitrate': '20M'
                },
                'tcp': {
                    'bitrate': '0'
                }
            }
        }
    ],
}
``` 

Does it make sense?

Aryamanz29 replied:
Yes, it would be better to have a single setting for `iperf check` :+1:

nemesisdesign reviewed (changes requested):
Great progress @Aryamanz29, see my comments below.

Aryamanz29 commented:
### `Iperf charts`
![charts1](https://user-images.githubusercontent.com/56113566/180446057-84befb47-80bc-4fdb-bd47-ba3cbb112349.gif)

nemesisdesign reviewed (changes requested):
I finally found what I was trying to say regarding the chart!

Try this:
https://codepen.io/nemesisdev/pen/vYRZgmz

You see, there's a null value. If you change that with 5 the result will be almost identical.

That's what I would like to do with these charts for IPERF, we will have a lot of missing values, it would be great to connect the gaps.

The visual result will be a lot better if we manage to avoid those gaps.

Aryamanz29 commented:
> I finally found what I was trying to say regarding the chart!
> 
> Try this: https://codepen.io/nemesisdev/pen/vYRZgmz
> 
> You see, there's a null value. If you change that with 5 the result will be almost identical.
> 
> That's what I would like to do with these charts for IPERF, we will have a lot of missing values, it would be great to connect the gaps.
> 
> The visual result will be a lot better if we manage to avoid those gaps.

With `fill(none)` in `influxdb query` we can achieve the same results, More info : https://docs.influxdata.com/influxdb/v1.8/query_language/explore-data/?t=Example+3%3A+fill%28none%29#fillfill_option

pandafy reviewed (commented):
@Aryamanz29 I tested the latest changes and the scale of  the x-axis on the graph looks too precise 

![Screenshot from 2022-07-25 23-37-46](https://user-images.githubusercontent.com/32094433/180845074-e23df0a5-ece0-4ec3-9423-f9f296cb6e88.png)

I got this with 1 day filter. The check was executed only twice.

pandafy reviewed (changes requested):
Good progress @Aryamanz29! We are close to completion of this PR. I have requested some minor changes below..

Aryamanz29 commented on README.rst:
```diff
@@ -803,6 +812,67 @@ Mobile Access Technology in use
+.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/datagram-loss.png
+  :align: center
+
+**Note:** Iperf charts uses ``connect_points=True`` in 
+`default chart configuration <#openwisp_monitoring_charts>`_ that joins it's individual chart data points.
```
I and @nemesisdesign had meet in which he suggested me to add this about `iperf charts`.

pandafy replied:
Okay, thank you!

Aryamanz29 commented on openwisp_monitoring/monitoring/static/monitoring/js/chart.js:
```diff
@@ -251,6 +261,10 @@
                 }
                 options.y.push(val);
+                // push only those 'x' points to options.x which has non null 'y' points
+                if(options._connectPoints){
+                    options.x.push(xValuesRaw[c]);
+                }
```
ploty `connectgaps` property only works for scatter charts. More info : https://plotly.com/python-api-reference/generated/plotly.graph_objects.Scatter.html

pandafy replied:
okay, thank you!

Aryamanz29 commented on README.rst:
```diff
@@ -334,6 +334,7 @@ Follow the setup instructions of `openwisp-controller
 .. code-block:: python
 
     from django.conf import settings
+    from celery.schedules import crontab
```
thanks! I will update :+1:

Aryamanz29 commented on README.rst:
```diff
@@ -939,6 +1009,214 @@ configuration status of a device changes, this ensures the check reacts
+#########################################################
+
+Follow the steps in `"How to configure push updates" section of the openwisp-controller documentation 
+<https://github.com/openwisp/openwisp-controller#how-to-configure-push-updates>`_ to allow SSH access to you device from OpenWISP.
```
Sure :+1:

Aryamanz29 commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,364 @@
+        """
+        return dc.connect()
+
+    def _deep_get(self, dictionary, keys, default=None):
```
Sure 👍

Aryamanz29 commented on openwisp_monitoring/monitoring/configuration.py:
```diff
@@ -542,6 +543,185 @@ def _get_access_tech():
+                },
+                'trace_order': ['total', 'received', 'sent'],
+                'description': _('Bitrate during Iperf3 test in TCP mode.'),
+                'summary_labels': [
+                    _('Total bitrate'),
+                    _('Sent bitrate'),
+                    _('Received bitrate'),
+                ],
```
thanks!

Aryamanz29 commented:
### `Iperf adaptive charts`

![Screenshot from 2022-07-29 18-23-27](https://user-images.githubusercontent.com/56113566/181764869-69e2c576-44b5-4f2c-bbf6-0b6f593fc226.png)

![Screenshot from 2022-07-29 18-20-39](https://user-images.githubusercontent.com/56113566/181764462-cc63a8fc-9fdc-40ce-a45f-bf4cddc3570c.png)

![Screenshot from 2022-07-29 18-20-27](https://user-images.githubusercontent.com/56113566/181764488-81a2c645-9994-4cb4-890e-3a6230711623.png)

![Screenshot from 2022-07-29 18-23-34](https://user-images.githubusercontent.com/56113566/181764816-bda11030-4d15-4cf2-9d73-95985ece32cc.png)

pandafy commented:
@Aryamanz29 I still get this locally 

![Screenshot from 2022-08-02 12-55-21](https://user-images.githubusercontent.com/32094433/182316755-b0fcf9e3-c058-45cb-9452-5a3d20c123c4.png)


See the time range on the x-axis. I think we should ensure that the x-axis common for all the charts.

nemesisdesign commented on openwisp_monitoring/monitoring/static/monitoring/js/chart.js:
```diff
@@ -355,8 +362,11 @@
-                if (unit == 'adaptive_bytes') {
-                    adaptiveFilterSummary(i, percircles, value);
+                if (unit.includes('adaptive_prefix')) {
+                    if(unit.includes('+')){
```
respect the spacing (I wish the linter warned about this!):

```
if (unit.includes('+')) {
`

Aryamanz29 replied:
Linter didn't warn 🤔 I'll check all my code manually :+1: 

![Screenshot from 2022-08-04 21-54-41](https://user-images.githubusercontent.com/56113566/182901953-77fbb3ee-110b-4660-ba2b-c2d5fdb60f2b.png)

nemesisdesign commented on openwisp_monitoring/monitoring/static/monitoring/js/chart.js:
```diff
@@ -271,15 +271,22 @@
                 yRawVal = data.traces[i][1];
             }
-            adaptiveFilterPoints(charts, layout, yRawVal);
+            if(data.connect_points){
```
same here and below

nemesisdesign commented on README.rst:
```diff
@@ -803,6 +811,67 @@ Mobile Access Technology in use
+
+**Transfer (TCP)**:
+
+.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/transfer-tcp.png
```
let's do those changes to the charts that we discussed (have bandwidth and transfer as simple lines with no fill and have 2 lines in each chart, one for TCP and one for UDP, then update the README here and screenshots.

nemesisdesign commented on README.rst:
```diff
@@ -1022,6 +1311,70 @@ validating custom parameters of a ``Check`` object.
++--------------+-------------------------------+
+
+This setting allows you to choose RSA public key path (relative to ``/root``)
+for `iperf check running with authentication <#configure-iperf-check-for-authentication>`_.
```
I don't understand, isn't this something that depends on a specific server? Why do we have a global setting?

Aryamanz29 replied:
Yes it should be defined in [`OPENWISP_IPERF_CHECK_CONFIG`](https://github.com/openwisp/openwisp-monitoring/tree/issue-385/iperf-check#openwisp_monitoring_iperf_check_config)

nemesisdesign commented on README.rst:
```diff
@@ -1603,6 +1956,13 @@ The ``AlertSettings`` of ``ping`` metric will by default use ``threshold`` and `
+which allows ``AlertSettings`` to use ``related_field`` as value to check ``threshold`` instead
+of default ``field_name`` key. A real world example of this can be seen in 
+`Iperf metric configuration 
+<https://github.com/openwisp/openwisp-monitoring/blob/issue-385/iperf-check/openwisp_monitoring/monitoring/configuration.py#L546-L700>`_,
```
definitely we cannot link files which are published in branches we are going to delete, please double check the rest of the document for similar problems, we can point to master here, even thought it won't work until published, but that's ok.

nemesisdesign commented on README.rst:
```diff
@@ -1912,7 +2272,7 @@ Management commands
 ~~~~~~~~~~~~~~
 
-This command will execute all the `available checks <#available-checks>`_ for all the devices.
+This command will execute all the `available checks <#available-checks>`_ for all the devices. 
```
this change is unnecessary, please revert it

nemesisdesign commented on openwisp_monitoring/check/classes/iperf.py:
```diff
@@ -0,0 +1,367 @@
+                'properties': {
+                    'bitrate': {
+                        'type': 'string',
+                        'default': '10M',
```
shall we default this to 30 MB? I think 10 is quite low.

Aryamanz29 replied:
Sure

nemesisdesign commented on README.rst:
```diff
@@ -939,6 +1018,216 @@ configuration status of a device changes, this ensures the check reacts
+
+It also supports tuning of various parameters.
+
+You can also change the parameters used for iperf checks (e.g. timing, port, username, password, rsa_publc_key etc) 
```
please break lines and maintain line length consistent with the rest of the document

nemesisdesign commented on openwisp_monitoring/monitoring/static/monitoring/js/chart.js:
```diff
@@ -196,7 +197,10 @@
-                        options.fill = "none";
+                        options.fill = data.fill;
+                    }
+                    if (options._connectPoints){
```
make sure spacing around parenthesis is consistent with the rest of the JS code

nemesisdesign commented on openwisp_monitoring/monitoring/static/monitoring/js/chart.js:
```diff
@@ -251,21 +264,32 @@
                 }
                 options.y.push(val);
+                // push only those 'x' points to options.x which has non null 'y' points
+                if(options._connectPoints){
```
same here

nemesisdesign commented on openwisp_monitoring/monitoring/static/monitoring/js/chart.js:
```diff
@@ -251,21 +264,32 @@
                 yRawVal = data.traces[i][1];
             }
-            adaptiveFilterPoints(charts, layout, yRawVal);
+            if(data.connect_points){
```
double check all blocks please

nemesisdesign commented on openwisp_monitoring/monitoring/base/models.py:
```diff
@@ -766,7 +792,15 @@ def _is_crossed_by(self, current_value, time=None, retention_policy=None):
+                        )
+                    )
+                else:
+                    results.append(self._value_crossed(point[self.metric.field_name]))
```
isn't the final return statment suppposed to be updated here??

```
return self._time_crossed(time) and value_crossed
```

Is not taking in consideration the related field.

What if value_crossed is False but the related value is crossed?

What about `if self.tolerance == 0:`???

So now I have some more basic quesitons:

1. Why was this kind of change, which is very delicate, introduced in this huge PR? Do we have to introduce this change or can we make the iperf check work without it for now? 

2. Looking at this code, I am pretty scared to merge it! I do not want to do this.
@Aryamanz29 can you extract these commits into a separate branch please? I want to merge the rest of the work but not this right now, I think it needs more careful thought.

Aryamanz29 replied:
1. Yes, we can make `iperf check` work without this feature (check will run without any `alertsettings`)
2. I'll remove `alert_on_related_field` feature from this PR as of now & will open a separate PR :+1:

nemesisdesign commented on openwisp_monitoring/monitoring/static/monitoring/js/chart.js:
```diff
@@ -235,6 +244,10 @@
                 // prepare data shown in chart on hover
                 if (val === null) {
+                    // filter 'y' null points from options.y
+                    if(options._connectPoints) {
```
same here

nemesisdesign reviewed (changes requested):
Good progress @Aryamanz29, please address my comments below.

Aryamanz29 commented on README.rst:
```diff
@@ -939,6 +1007,221 @@ configuration status of a device changes, this ensures the check reacts
+               """
+           ),
+           # relative to /root directory
+           'rsa_public_key_path' : '/tmp/iperf-public-key.pem',
```
This is the location where the `client (ie. openwrt device)` stores it's `rsa_public_key` for `iperf check`. By default, it is `/tmp/iperf-rsa-public.pem` and this path is required for iperf3 auth, but we can also change this using [OPENWISP_IPERF_CHECK_CONFIG](https://github.com/openwisp/openwisp-monitoring/tree/issue-385/iperf-check#openwisp_monitoring_iperf_check_config). More info : https://github.com/openwisp/openwisp-monitoring/issues/414#issuecomment-1189408176

Aryamanz29 replied:
Iperf command basically doing  `f'echo "{rsa_public_key}" > {rsa_public_key_path}` and after completion of check removing key from the device `rm {rsa_public_key_path}`

nemesisdesign replied:
`/tmp/iperf-public-key.pem` does not need to be configurable at this level, there's no usefulness in this, just set `/tmp/iperf-public-key.pem` to a variable and use the variable throughout the rest of the implementation.

Aryamanz29 replied:
Ok :+1:

nemesisdesign reviewed (changes requested):
I pushed some improvements to the README, please double check!

Please also do the following changes:

- add a reference (with the proper link) to the Iperf feature in the `Available Features` section
- in the section which lists the metrics, when Iperf is referenced, please provide a link to the `#iperf-check-usage-instructions` section, eg: "For more info on how to configure and use Iperf, please refer to ....".

pandafy reviewed (approved):
LGTM 👍🏼 
Let's merge this!
